### ISSUE TITLE
Support "uniform texel buffer" and "storage texel buffer" ?

### ISSUE THREAD

**Issue by dneto0** [gpuweb/gpuweb#162]
* Uniform texel buffers permit loads on a buffer view, with value-conversion similar to image reads.  See Vulkan's [13.1.5 Uniform Texel Buffer](https://www.khronos.org/registry/vulkan/specs/1.1/html/vkspec.html#descriptorsets-uniformtexelbuffer)
   * If supported, then enable the SampledBuffer capability in the WebGPU env spec for SPIR-V

* Storage texel buffers permit loads and stores on a buffer view, with value-conversion similar to image reads and writes.  See Vulkan's [13.1.6 Storage Texel Buffer](https://www.khronos.org/registry/vulkan/specs/1.1/html/vkspec.html#descriptorsets-storagetexelbuffer)
   * If supported, then enable the ImageBuffer capability in the WebGPU env spec for SPIR-V

I don't have an opinion on whether these should be supported by WebGPU, and I didn't see an investigation related to this.

**Comment by devshgraphicsprogramming** [comment]
If they're not supported in WebGPU then the developer will have to emulate them with SSBOs, except that the value-conversion (i.e. RGB9E5 to vec4) will be much slower as opposed to hardware.

Secondly there is an issue with the lack of support for Uniform Texel Buffers (TBO in OpenGL parlance), because in Vulkan an SSBO has to conform to the std430 layout, so supporting formats of less than 4 bytes per "pixel" (or rather sample) will be very painful.

This will be a bit hard to investigate as Vulkan Hardware database has no information about supported formats for Uniform and Storage Texel Buffers as well as the shader stages that support them.

**Comment by kvark** [comment]
@devshgraphicsprogramming good point.
Btw, the stages at least seem to be well defined by the spec (and thus can be derived from the features in vulkandb).

Uniform texel buffers are exposed to all stages:
> Load operations from uniform texel buffers are supported in all shader stages for image formats which report support for the VK_FORMAT_FEATURE_UNIFORM_TEXEL_BUFFER_BIT feature bit via vkGetPhysicalDeviceFormatProperties in VkFormatProperties::bufferFeatures.

Storage texel buffers are exposed to the same stages as SSBO/UAV in general:
> When the fragmentStoresAndAtomics feature is enabled, stores and atomic operations are also supported for storage texel buffers in fragment shaders with the same set of texel buffer formats as supported in compute shaders. When the vertexPipelineStoresAndAtomics feature is enabled, stores and atomic operations are also supported in vertex, tessellation, and geometry shaders with the same set of texel buffer formats as supported in compute shaders.

**Comment by devshgraphicsprogramming** [comment]
Yeah so emulating R8,R8G8, R16, etc. with SSBOs would be a pain.

Basically always read the 32bits, then extract the bit range and optionally normalize. This would be a big perf-loss against native Vulkan or OpenGL.

**Comment by dneto0** [comment]
> Hopefully not for much longer! https://renderdoc.org/vkspec_chunked/chap41.html#VK_EXT_scalar_block_layout

Yes, that's the bleeding edge of support.  Hopefully it's the last time Vulkan has to relax the layout rules!

The flip side is whether a big enough installed base can get that extension in time for WebGPU rollout.

Perf impact: Yes, unfavourably aligned vector accesses will be slower on some hardware, whether or not that extension is supported.  Tradeoff is whether the user has to do the load-32-bits-then-unpack vs. the implementation.  At least if the implementation has the responsibility, they have the opportunity of doing it better; and your shader code is more clear.

**Comment by Kangz** [comment]
Tentatively closing. IIRC the group decided to not have these features and nobody asked for them (SSBOs mostly just better though there's some discussion about that in #297)

**Comment by kvark** [comment]
@Kangz this may become important for OpenGL backends. There are GL platforms that plain refuse to expose any SSBO stuff, but they'd be totally usable through TBOs.

**Comment by Kangz** [comment]
Do these GL versions support compute shaders? I thought compute shaders always came with SSBOs.

**Comment by kvark** [comment]
Here is some light reading - https://www.raspberrypi.org/forums/viewtopic.php?t=271863
It talks about GLES-3.1, so it must have the compute stuff.

**Comment by Kangz** [comment]
This is only for vertex shaders, so not quite as bad but this is a real constraint when targeting ES 3.1. I think it might be possible to magically transform VS using SSBO into a VS using a TBO (similar to the ByteAddressBuffer transform) and then back the TBO with the storage buffer. This would work for readonly-storage-buffer at least (there is no way to implement RW storage buffers in ES 3.1 AFAIK)

**Comment by magcius** [comment]
You can kind of emulate SSBOs using imageLoadStore, which I believe should be in ES 3.1

**Comment by dneto0** [comment]
The Chrome team is getting multiple requests for texel buffers.

Motivations include:

* Ability to read and write to individual scalars from large arrays or rectangular grids of data. E.g. read and write 8bit bytes, on hardware that is never going to support general access to scalars narrower than 32bits.
* Do the above, for data living in a GPUBuffer.   In contrast, storage textures live in a GPUTexture). 
* Exploit performance differences between memory types.   This is GPU-dependent, but can be quite significant, particularly depending on access pattern.
      *  @sebbbi has a project to benchmark *load* speeds for various memory types using various access patterns. https://github.com/sebbbi/perftest/blob/master/README.md 
      * I would like to revisit those experiments and also investigate write performance.
* Reducing friction when porting applications.

**Comment by dneto0** [comment]
Additionally, storage texture coordinates are limited.  
WebGPU mandates supporting 1D and 2D coordinates up to 8192.

Texel buffers are not limited in this way. They would be limited by the size of the underlying GPUBuffer.

**Comment by dneto0** [comment]
Moved to Milestone 2, per discussion in WebGPU API meeting 2023-09-27

**Comment by devshgraphicsprogramming** [comment]
> > because in Vulkan an SSBO has to conform to the std430 layout
> 
> Hopefully not for much longer! https://renderdoc.org/vkspec_chunked/chap41.html#VK_EXT_scalar_block_layout

can confirm that finally in 2023 all relevant (still supported) desktop platforms support scalar block layout, as for mobile... I cannot say

**Comment by kdashg** [comment]
<details><summary><a href="https://github.com/gpuweb/gpuweb/wiki/WGSL-2023%E2%80%9012%E2%80%9005-Minutes#add-texel-buffers-162">WGSL 2023-12-05 Minutes</a></summary>

* AB: We have partner requests for this. It comes from various places. Fills gap on older hardware that doesn't have smaller data type support. Very common in older D3D to get read write storage of f16, say, without needing full f16 type support. Widely used.
* DN: this lets you treat GPUBuffer memory as a big array of textures. Does not have coordinate size limitations. But use the texel format conversion hardware on the data path into/out of the shader.
* JB: Learning about this feature. Don’t express an opinion.
* AB: Certain use cases can be satisfied with f16. There are other use cases to help porting existing code. There’s also a reach aspect. E.g. D3D doesn’t have f16 until a later SM 6.x but prior APIs have texel buffers which is how you end up covering for it.
* JB: Sounds like some folks really need it but many don’t care.
* KG: Estimates on how much work this is to add?
* DN: Mostly just plumbing, so should be easy enough.
* JB: Ok! Mozilla is neutral. M1 is tolerable but not preferred.
* **->M1**

</details>

**Comment by jimblandy** [comment]
It would be nice to know how widely this feature is supported on our target hardware, to decide whether this would be a feature or not optional.

Concretely, what would this look like in the API and in WGSL?

**Comment by kdashg** [comment]
<details><summary><a href="https://github.com/gpuweb/gpuweb/wiki/WGSL-2024‐05‐14-Minutes#support-uniform-texel-buffer-and-storage-texel-buffer---issue-162--gpuwebgpuweb--github">WGSL 2024-05-14 Minutes</a></summary>

* KG: This is showing up now, why?  Kai was going through old issues and adding tags.  This touches on WGSL. We last talked about it in WGSL in Dec 2023.  At the time we put it in M1.  As we talked in Mozilla, it would be nice to know how widely supported it is; is it optional, and understand the API side implications. On the WGSL side is there another address space.
* AB: It’s widely supported.  We’re ok moving this to M2. It would be a new texture type, with storage in the ‘handle’ space.
* KG: Seems an API proposal stage.
* AB: We have an internal doc we can expose for the investigation.
* DN: This *is* requested by partners, has real demand. But, this is lower in priority than e.g. subgroups, and we don’t expect to implement these new until much later this year.
* MM: is this implementable on metal?
* AB: i think it's just texture_buffer in metal
* MW: All Metal devices should be able to support this.
* DN: I updated to Milestone 2.

</details>

**Comment by litherum** [comment]
> The Chrome team is getting multiple requests for texel buffers.

Is a texel buffer a buffer that is backed by texture data, or a texture that is backed by buffer data?

If I'm not mistaken, Metal only [supports](https://developer.apple.com/documentation/metal/mtlbuffer/1613852-newtexturewithdescriptor) the latter.

**Comment by mwyrzykowski** [comment]
That's right, Metal texture buffers, `MTLTextureTypeTextureBuffer`, are 1D textures which don't support mipmaps or texture arrays.

There are also buffer backed 2D textures which can be created from [-[MTLBuffer newTextureWithDescriptor:]](https://developer.apple.com/documentation/metal/mtlbuffer/1613852-newtexturewithdescriptor)

In both cases, no support for mipmaps, array length is always 1, sample count is always 1, no support for compressed formats, maybe I am forgetting something.

**Comment by litherum** [comment]
> That's right, Metal texture buffers, `MTLTextureTypeTextureBuffer`, are 1D textures which don't support mipmaps or texture arrays.

BTW, from the [MSL spec](https://developer.apple.com/metal/Metal-Shading-Language-Specification.pdf) in section 2.9.1 Texture Buffers, it says:

> However, you cannot sample a texture buffer.

**Comment by litherum** [comment]
> Ability to read and write to individual scalars from large arrays or rectangular grids of data.

Re: "rectangular grids," does Vulkan support that? It appears the way to create a uniform texel buffer or a storage texel buffer is by creating a `VkBufferView`, but [`VkBufferViewCreateInfo`](https://registry.khronos.org/vulkan/specs/1.3-khr-extensions/html/chap12.html#VkBufferViewCreateInfo) doesn't include any fields regarding the Y dimension. It appears only capable of creating 1-dimensional textures.

**Comment by litherum** [comment]
I think the use case of "be able to have different threads write to adjacent bytes without racing" is a compelling use case.

**Comment by litherum** [comment]
> Exploit performance differences between memory types.

I was interested in characterizing this, so I wrote a [little benchmark](https://github.com/gpuweb/gpuweb/files/15478665/BufferTexturePerformance.zip) to see what the performance difference was in Metal.

The benchmark is straightforward:
- Sum 33M floats by reading them sequentially from a texture buffer
- Sum the same 33M floats by reading them in random order from a texture buffer. Here, "random order" is a trick I learned from [Ray Tracing Gems](https://www.realtimerendering.com/raytracinggems/rtg/index.html): if you simply reverse the bits in the index, your reads will be in effectively random order. Reversing the bits of a number can be done in [just a few instructions](https://stackoverflow.com/a/2602885/125848)
- Sum the same 33M floats by reading them sequentially from their underlying buffer
- Sum the same 33M floats by reading them in random order from their underlying buffer

Here are the performance results:

<img src="https://github.com/gpuweb/gpuweb/assets/918903/018c6b1d-bd4d-4911-b329-d076126c03e4">

Surprisingly, this shows that reading from the raw buffer is faster than reading from a texture buffer. However, I don't think this should sink the feature, because there are [use cases](https://github.com/gpuweb/gpuweb/issues/162#issuecomment-2136070649) beyond simply read performance in shaders.

**Comment by litherum** [comment]
Another interesting tidbit: in Vulkan, not every pixel format can be used as a uniform/storage texel buffer. And, the spec doesn't list which formats can be used (and which cannot be used). Instead, the [spec says](https://registry.khronos.org/vulkan/specs/1.3-extensions/html/vkspec.html#resources-buffer-view-format-features):

> If Vulkan 1.3 is supported or the [VK_KHR_format_feature_flags2](https://registry.khronos.org/vulkan/specs/1.3-extensions/html/vkspec.html#VK_KHR_format_feature_flags2) extension is supported, then the buffer view’s set of format features is the value of [VkFormatProperties3](https://registry.khronos.org/vulkan/specs/1.3-extensions/html/vkspec.html#VkFormatProperties3)::bufferFeatures found by calling [vkGetPhysicalDeviceFormatProperties2](https://registry.khronos.org/vulkan/specs/1.3-extensions/html/vkspec.html#vkGetPhysicalDeviceFormatProperties2) on the same format as [VkBufferViewCreateInfo](https://registry.khronos.org/vulkan/specs/1.3-extensions/html/vkspec.html#VkBufferViewCreateInfo)::format.

So, you have to ask _at runtime_ whether or not the format is compatible with uniform/storage texel buffers, and different devices can return arbitrarily different answers for any given format.

(Aside: I have no idea how Vulkan native app developers could possibly use such an API effectively. What are you supposed to do if the device you're on just happens to not support the codepath you wrote? Write a codepath for literally every format? Fall back to buffers, and hope the device has [`VK_KHR_8bit_storage`](https://registry.khronos.org/vulkan/specs/1.3-extensions/man/html/VK_KHR_8bit_storage.html)?)

**Comment by sebbbi** [comment]
In DirectX, read-only buffer (Buffer<T>) is supported for all types. Texel buffers all use opaque descriptor in DirectX. The shader doesn't know the buffer type. Sampler hardware does the type conversion.  

DirectX RWBuffer<T> has format limitations. And there's multiple tiers of "typed UAV load" support:
https://learn.microsoft.com/en-us/windows/win32/direct3d12/typed-unordered-access-view-loads

RWBuffers are also bound using different UAV binding model. Read-only buffers are bound as SRV (shader resource view).

**Comment by sebbbi** [comment]
Typed UAV Load (extended formats) support can be found in the following table:
https://en.wikipedia.org/wiki/Feature_levels_in_Direct3D

Nvidia Fermi (GTX 500) and Kepler (GTX 600 and 700) don't suppor it and Intel Gen 7.5 (Haswell) and Gen 8 (Broadwell) don't support it. All other DX12 capable hardware supports it.

**Comment by Kangz** [comment]
That it is an optional feature in D3D12 and Vulkan mean that this will need to be an extension. To find which format support that feature in Vulkan (and even D3D12), we can volunteer to add some stat gathering in Chromium to see what several options would give in terms of reach in the wild.

**Comment by jimblandy** [comment]
<details>
<summary>minutes from WGSL committee meeting 2026-9-15</summary>
<ul>
<li>JB: I think GitHub automation incorrectly moved this to “In Progress”, but I don’t remember us prioritizing texel buffers, and the texel buffer proposal didn’t make the top three list, so I moved this to late M2.</li>
<li>Yes, late M2 is okay for this</li>
<li>JB: [added after meeting] In triage discussion, we decided that since we were unlikely to work on this, let alone land it, before November, it didn’t belong in M2.</li>
</ul>
</details>
